doc: update packages documentation for Node.js 12 EOL (#43375)
nodejs/node: doc/api/packages.md

## Package entry points

In a package's `package.json` file, two fields can define entry points for a
package: [`"main"`][] and [`"exports"`][]. Both fields apply to both ES module
and CommonJS module entry points.

The [`"main"`][] field is supported in all versions of Node.js, but its
capabilities are limited: it only defines the main entry point of the package.

The [`"exports"`][] provides a modern alternative to [`"main"`][] allowing
multiple entry points to be defined, conditional entry resolution support
between environments, and **preventing any other entry points besides those
defined in [`"exports"`][]**. This encapsulation allows module authors to
clearly define the public interface for their package.

> For new packages supporting Node.js 12.+ the [`"exports"`][] field is

Reviewer: ```suggestion
For new packages supporting Node.js 12+ the [`"exports"`][] field is
```
or
```suggestion
For new packages supporting Node.js 12.x+ the [`"exports"`][] field is
```

    "./feature/index": "./feature/index.js",
    "./feature/index.js": "./feature/index.js",
    "./package.json": "./package.json"
  }
}
```

Alternatively a project could choose to export entire folders both with and
without extensioned subpaths using export patterns:

```json
{
  "name": "my-package",
  "exports": {
    ".": "./lib/index.js",
    "./lib": "./lib/index.js",
    "./lib/*": "./lib/*.js",
    "./lib/*.js": "./lib/*.js",
    "./feature": "./feature/index.js",
    "./feature/*": "./feature/*.js",
    "./feature/*.js": "./feature/*.js",
    "./package.json": "./package.json"
  }
}
```

With the above providing backwards-compatibility for any minor package versions,
a future major change for the package can then properly restrict the exports
to only the specific feature exports exposed:

```json
{
  "name": "my-package",
  "exports": {
    ".": "./lib/index.js",
    "./feature/*.js": "./feature/*.js",
    "./feature/internal/*": null
  }
}
```

Where it is then recommended to pick one of supporting extensioned or
unextensioned subpaths for consistent usage. In the above example the
extensioned form is used (ie via `import "my-package/feature/feat.js"`), which

> can be useful for better [interoperability with import maps][].

Reviewer 2: This is much better, but interop remains the same. It should be up to consumers to decide if the ergonomics of extensionless imports are worth a slightly larger import map, or not.

Author: Packages that define only unextensioned subpaths cannot be consumed any other way and so the consumer still has no choice here but to use what the package picks.

}
```

Alternatively a project could choose to export entire folders using export
patterns:

```json
{
  "name": "my-mod",
  "exports": {
    ".": "./lib/index.js",
    "./lib": "./lib/index.js",

>     "./lib/*": "./lib/*",

Reviewer 2: Entry points shouldn’t contain extensions; won’t this change mean consumers have to provide the .js?

Author: See bullet point two - one of the changes in this PR is to encourage explicit entry extensions, since it enables compressible import maps via `./` mappings in the map, which would otherwise require explicit mappings.

It's certainly something that may have arguments either way though, and if there is resistance to the change I could remove this aspect of the PR, although it is one of the motivating reasons in trying to guide the ecosystem towards a place where import maps don't end up bloated.

Reviewer 3: One of the things that came up in the collaborator summit is that Node needs first-class support for import maps, so +1 on starting to move us in that direction.

Reviewer 2: That contradicts the longstanding ecosystem preference, and i don’t think it belongs in node’s docs.

Import maps from node_modules are useless unless you a) serve node_modules, a horrifically bad practice, and b) have a 100% ESM tree, which virtually nobody does, so i don’t think it’s valuable to artificially push the ecosystem in that direction.

Reviewer: Having `"exports"` compatible with import maps lets users generate an import maps that use e.g. `unpkg.com` so you could start using your JS app (with Deno, a web browser, Node.js once it has import maps support, etc.) without needing a node_modules folder. My interpretation of things is not that Node.js is artificially pushing the ecosystem to ESM, but rather the ecosystem is organically pushing Node.js to ESM and interoperability with other tools of the ecosystem, and import maps are part of this.

Alternatively, we could give an example of an import maps, and illustrate how export pattern, while supported by Node.js, may not be desirable because it makes the process of generating an import maps for your module more complicated and error-prone. That would let users choose one or the other depending on wether import maps support matters to their use-case.

So given this `package.json`, we can use the following import maps:

```json
{
  "imports": {
    "my-mod": "https://unpkg.com/my-mod/lib/index.js",
    "my-mod/lib": "https://unpkg.com/my-mod/lib/index.js",
    "my-mod/lib/": "https://unpkg.com/my-mod/lib/"
  }
}
```

If the `package.json` uses export pattern, it would look something like:

```json
{
  "imports": {
    "my-mod": "https://unpkg.com/my-mod/lib/index.js",
    "my-mod/lib": "https://unpkg.com/my-mod/lib/index.js",
    "my-mod/lib/index": "https://unpkg.com/my-mod/lib/index.js",
    "my-mod/lib/my-mod": "https://unpkg.com/my-mod/lib/my-mod.js",
     …
    "my-mod/lib/yetAnotherFile": "https://unpkg.com/my-mod/lib/yetAnotherFile.js",
  }
}
```

BTW, `my-mod` is not a very good package name, it blurs the line between a package and a module which is very confusing.

Reviewer 2: Any such import map generating tool can do so whether extensions or used or not. It is a fallacy that this has any meaningful impact on import map generation imo.

Author: The second example @aduh95 provides above clearly has more entries than the first. If we had `*` patterns in import maps, sure it wouldn't be an issue, but import maps have web compat at this point such that their matching syntax is locked and unlikely to change.

Reviewer 2: Are they shipped in a non chrome browser?

Author: Yes, under a flag in Firefox.

Reviewer 3: +1 on renaming `my-mod`; maybe `package`? Also the use of `main.js` might be confusing since you also discuss `"main"`; maybe rename that to `entry.js`?

Author: I've renamed `my-mod` to `my-package` and replaced usage of `main.m?js` with `index.m?js`, let me know how that looks.

Reviewer 2: "under a flag" isn't part of the web, so that means they're just shipping in chrome.

Reviewer 3: Besides the web, import maps are a core feature of Deno: https://deno.land/manual/linking_to_external_code/import_maps. We’re also trying to improve interoperability between server-side JavaScript runtimes.

Reviewer 2: Sure, that’s all fine, extensionless entry points however remain dominant, idiomatic, and supported, so I’m not sure why “a feature node doesn’t support” would have more weight than that?

Author: Because the example here is explicitly about supporting backwards-compatibility, I've updated the example to use the pattern of exporting support for exports both with and without extensions. This should resolve this concern I believe.

}
```

Alternatively a project could choose to export entire folders both with and
without extensioned subpaths using export patterns:

```json
{
  "name": "my-package",
  "exports": {
    ".": "./lib/index.js",
    "./lib": "./lib/index.js",
    "./lib/*": "./lib/*.js",
    "./lib/*.js": "./lib/*.js",
    "./feature": "./feature/index.js",
    "./feature/*": "./feature/*.js",
    "./feature/*.js": "./feature/*.js",
    "./package.json": "./package.json"
  }
}
```

While the above supports good backwards-compatibility for package upgrades,
a future major change for the package can then properly restrict the exports
to only the specific feature exports exposed:

```json
{
  "name": "my-package",
  "exports": {
    ".": "./lib/index.js",
    "./feature/*.js": "./feature/*.js",
    "./feature/internal/*": null
  }
}

> ```

Author: I added this new section to note how the package can eventually be upgraded on a major. PTAL!

Reviewer 3: Looks good to me. Do we want to include that we’re recommending `*.js` in order to produce non-bloated import maps, or is that better mentioned somewhere else?

Along with the first mention of import maps we could maybe also say that Node intends to support import maps in the future. It’s on the [loaders roadmap](https://github.com/nodejs/loaders), and putting it here in the docs would make it yet more official.

Author: I've added a note below and also included a dedicated section on import maps compatibility, linking to that.

absolute subpath of the package such as
`require('/path/to/node_modules/pkg/subpath.js')` will still load `subpath.js`.

> For packages supporting Node.js < 12.7.0 it is necessary to include the `"main"`

Reviewer: More than old versions of Node.js, there are tools (e.g. bundlers) that don't support `"exports"`, maybe we should just say "to maximize backward-compatibility"?

Author: Yeah I can definitely call out other integration workflows further.

Author: I've clarified this section, let me know if that works better.

absolute subpath of the package such as
`require('/path/to/node_modules/pkg/subpath.js')` will still load `subpath.js`.

> All modern JS build tools support the `"exports"` field, but if necessary for
> compatibility with specific tooling or Node.js versions < 12.7.0, the `"main"`
> field can be included alongside `"exports"`:

Reviewer 3: There are lots of projects stuck on non-modern build tools, like Webpack 4, which are hard to upgrade.

```suggestion
All supported versions of Node.js (since 12.7.0) and modern build tools
support the `"exports"` field. If a project is using an outdated version of
Node.js or a build tool, compatibility can be achieved by including the
`"main"` field alongside `"exports"`:
```

Author: Added with some adjustments.

absolute subpath of the package such as
`require('/path/to/node_modules/pkg/subpath.js')` will still load `subpath.js`.

All supported versions of Node.js (since 12.7.0) and modern build tools
support the `"exports"` field. For projects using an older version of Node.js
or a related build tool, compatibility can be achieved by including the `"main"`

> field alongside `"exports"`:

Reviewer 2: this should note that they also need to point to the same path, or at least something CJS will successfully resolve.

Author: Added.

```json
{
  "exports": {
    ".": "./index.js",

>     "./submodule.js": "./src/submodule.js"

Reviewer 2: ```suggestion
    "./submodule": "./src/submodule.js"
```

Reviewer 2: i still think these examples shouldn't be changed to _add_ the extension.

Reviewer: I don’t really understand what’s motivating this change either, if anything that adds more bloat to the import maps – I mean it adds 3 bytes; jokes aside, it could give the reader the impression that using an extension is required when defining an export path, and/or confuse needlessly the package consumers on why the is no `submodule.js` file at the trot of the package – whereas extensionless makes a good job signaling we’re not dealing with actual paths imo. 
(Same thing for the change `main.js` -> `index.js`, I don’t think it’s necessary)

That being said, I don’t feel strongly about this, so feel free to disregard.

  }
}
```

Now only the defined subpath in [`"exports"`][] can be imported by a consumer:

```js

> import submodule from 'es-module-package/submodule.js';

Reviewer 2: ```suggestion
import submodule from 'es-module-package/submodule';
```

// Throws ERR_PACKAGE_PATH_NOT_EXPORTED
```

Even though subpaths provide an aribtrary string mapping to the package
interface, it is recommended (but not required) to use explicit file extensions
when defining package subpaths so that package consumers write
`import 'pkg/subpath.js'` instead of `import 'pkg/subpath'` as this simplifies

> interop with other ecosystem tooling patterns such as when using import maps.

Reviewer 3: ```suggestion
interoperability with other ecosystem tooling patterns such as when using import maps.
```

// Throws ERR_PACKAGE_PATH_NOT_EXPORTED
```

Even though subpaths provide an aribtrary string mapping to the package
interface, it is recommended (but not required) to use explicit file extensions
when defining package subpaths so that package consumers write
`import 'pkg/subpath.js'` instead of `import 'pkg/subpath'` as this simplifies

> interoperability with other ecosystem tooling patterns such as when using import
> maps. This also mirrors the requirement of using [the full specifier path][] in

Reviewer 3: ```suggestion
interoperability with [import maps](https://github.com/WICG/import-maps).
Import maps are a cross-platform standard for module resolution, already in use
by some browsers, by other server-side JavaScript runtimes and by build tools.
This also mirrors the requirement of using [the full specifier path][] in
```

Author: Also added with some rewordings.

// Throws ERR_PACKAGE_PATH_NOT_EXPORTED
```

> #### Extensioned v Extensionless Subpaths

Author: I've renamed the title to `"Extensions in Subpaths"` rather than using competitive language!

Package authors should provide either extensioned (`import 'pkg/subpath.js'`) or
extensionless (`import 'pkg/subpath'`) subpaths in their exports. This ensures
that there is only one subpath for each exported module so that all dependents
import the same consistent specifier, keeping the package contract clear for
consumers and simplifying package subpath completions.

Traditionally, packages tended to use the extensionless style, which has the
benefits of readability and of masking the true path of the file within the
package.

With [import maps][] now providing a standard for package resolution in browsers

> and other server-side runtimes, using the extensionless style can result in

Reviewer: nit 
```suggestion
and other JavaScript runtimes, using the extensionless style can result in
```

// Throws ERR_PACKAGE_PATH_NOT_EXPORTED
```

> Even though subpaths provide an aribtrary string mapping to the package

Reviewer 2: ```suggestion
Even though subpaths provide an arbitrary string mapping to the package
```

interface, it is recommended (but not required) to use explicit file extensions
when defining package subpaths so that package consumers write
`import 'pkg/subpath.js'` instead of `import 'pkg/subpath'` as this simplifies

> interoperability with [import maps][] and also mirrors the requirement of using

Reviewer 2: i don't think this is an accurate claim. Compatibility with import maps is identical with or without extensions.

Author: I've added a new dedicated section on import maps compatibility.

[the full specifier path][] in relative and absolute import specifiers. Import
maps are a cross-platform standard for module resolution, already in use by some

> browsers, server-side JavaScript runtimes and build tools.

Reviewer 2: ```suggestion
browsers, server-side JavaScript runtimes, and build tools.
```

// Throws ERR_PACKAGE_PATH_NOT_EXPORTED
```

#### Extensioned v Extensionless Subpaths

It is recommended to pick either one of providing extensioned or unextensioned
subpaths for a package to ensure consistent usage. This keeps the package
contract clear for consumers, simplifies package subpath completions, and
ensures all dependent packages import with the same specifier mappings.

Ultimately, it is the choice of the package author which form of subpath to
support - extensionless, like `import 'pkg/subpath'`, or extensioned, like
`import 'pkg/subpath.js'` per the exports example above. Both conventions are
used in the Node.js ecosystem.

> For packages where interoperability with [import maps][] is desired, using

Reviewer 3: All package authors should desire interoperability with import maps. Even if Node never supports the feature (but it will) there are still things like Deno's Node compatibility mode. We should encourage whatever's the best approach for compatibility, both with the other runtimes of today and with Node's likely future.

Reviewer 2: I think that's why this phrasing is inaccurate - extensionless and extensioned are both identically interoperable with import maps. What package authors are likely to not all prioritize is minimizing the size of an import map, since "ergonomics" and "aesthetics" are often valued more than "bytes over the wire".

// Throws ERR_PACKAGE_PATH_NOT_EXPORTED
```

#### Interoperability with Import Maps

[Import maps][] are a separate cross-platform standard for module resolution,
already in use by some browsers, server-side JavaScript runtimes, and build

> tools, although not currently implemented or supported in Node.js.

Reviewer 3: ```suggestion
tools, although not currently supported by Node.js. We plan to implement
import maps support in Node.js in the future.
```

// Throws ERR_PACKAGE_PATH_NOT_EXPORTED
```

#### Interoperability with Import Maps

[Import maps][] are a separate cross-platform standard for module resolution,
already in use by some browsers, server-side JavaScript runtimes, and build
tools. Although not currently implemented in Node.js, there are plans to provide
import maps support in Node.js in future.

[Import maps][] are able to provide explicit specifier mappings for packages
and package subpaths, and therefore all mappings defined by Node.js package
exports and subpath exports can be mapped by a corresponding import map. In
addition they support [folder mappings][] but not pattern mappings like
Node.js supports.

> Where interoperability with import maps is desired, it is recommended to use

Reviewer 4: It's maybe too late to mention, but a canonical response [on stack overflow](https://stackoverflow.com/questions/2170745/what-are-the-pros-for-using-extension-less-urls) is to prefer extensionless URLs, per [this style article](https://www.w3.org/Provider/Style/URI) from the W3C. All of its logic certainly applies equally to module specifiers, which certainly could be resolved extensionlessly _without_ import maps with a correctly configured webserver. The inconsistent recommendations between what node maintainers are asking for and what the web community has published in the past are still weird, and still force consumers to be exposed to what should be package-internal format metadata, imo. Especially given export maps are a location you explicitly have the ability to hide that metadata.

Reviewer 3: Those are referring to URLs for public consumption, like a URL for sharing an article. That’s a case where you want a URL to be human-readable, whereas the URLs from an import map are only read by machines (or developers writing code).

There’s no reason a package author can’t publish _both_ the extensionless and the extensioned maps for a particular export. The point of this PR is to encourage the latter, as it solves a specific problem of bloated import maps, which lead to poorer performance. We’re taking an opinionated stance that this performance concern outweighs the prettiness that some perceive of extensioness imports in their code. This will become more important once Node adds its own support for import maps; we want to encourage the performant pattern as the recommended approach.

Reviewer 4: > Those are referring to URLs for public consumption, like a URL for sharing an article.

Package specifiers _are_ very widely shared URL fragments, often publicly consumed by other developers.

> We’re taking an opinionated stance that this performance concern outweighs the prettiness that some perceive of extensioness imports in their code.

You could equally take the stance that _hiding_ the extension, and thus changes in the underlying format compatible with the same import method, is preferable. And that has more tangible benefits for the majority of node users (by helping keep specifiers stable across changes in the underlying package), rather than hypothetical import map size.

> we want to encourage the performant pattern as the recommended approach

"Performant pattern" is an oxymoron with import maps. It is unambiguously less performant to use import maps than not - if performance is a goal, you should be avoiding import maps like the plague and ideally inlining all your specifiers that may have initially relied on them to exact paths. Import maps are a development time aid, not a production tool - if they're large, it should not matter.

Author: Reducing package complexity takes an incredible amount of work and commitment, and @weswigham you've done a huge and impressive job here yourself. It's unfortunate to hear there are so many catches in these workflows, although TypeScript has that other surprise of using a `.js` extension to load a `.ts` file which has always been incredibly surprising to me as well. Not sure you can blame the friction of the lack of intuitiveness on that one entirely on Node.js, especially when TypeScript is actively creating an interop barrier between Node.js and Deno applications in doing that.

We're all just trying our best to work towards having clear package conventions that can enable package authors to achieve their use cases while trying not to blindly lead the entire ecosystem off towards a complexity cliff in the future. We're actively thinking of interops and trying to mitigate them. To be honest, if more work like this had been done in the past we wouldn't be in the mess we are in today. It's unfortunate that those of us attempting to avoid the future gotachas are the ones getting the flak for causing inconvenience by attempting to work through conventions against different use cases and come up with ones that work well between different tools and approaches.

> Explaining that the import doesn't even actually refer to a file, per sey, yet still has an extension for some reason, is going to be a nice nuance to repeatedly explain.

Have there been any major issues in say exports completions work here? Surely the package author is entirely in charge of this by design? Node.js has always been this way in permitting both forms and continues to be. Package authors are free to choose. This is just an interop guidance section.

> It's going to be really cool when the pkg/file.js entrypoint has to be mapped to null in a deno condition in favor of a pkg/file.ts for no good reason other than someone saying it's a "best practice". It'll be a really cool and very legit pattern that definitely delivers value to users. :(

This is actually a really cool concept! It does actually seem to make sense to have that null map but rather for the Node.js condition not the Deno one:

```json
{
  "exports": {
    "./features/*.ts": {
      "node": null,
      "deno": "./src/*.ts"
    },
    "./features/*.js": {
      "deno": "./src/*.ts",
      "node": "./lib/*.js"
    }
  }
}
```

The above would then generate optimized import maps for Node.js, browsers and Deno applications for the package against the package exports, while avoiding the dual package hazard - but you don't want a null map for the js in Deno since you likely want to map the `.js` extensions to `.ts` extensions for compat between packages.

I frankly think it's amazing and a good thing that exports is able to achieve unforseen usage scenarios like the above. Package consumers don't need to know these details though, this is low level stuff for package authors to dip into as needed only - I really don't see how the cognitive overhead would burden the beginner user in any way.

> a canonical response [on stack overflow](https://stackoverflow.com/questions/2170745/what-are-the-pros-for-using-extension-less-urls) is to prefer extensionless URLs

Even with servers that handle the mapping you still have to choose between supporting `pkg/x.js` and `pkg/x`. If the server resolves `http://localhost/node_modules/pkg/x` into the file at `/pkg/x.js` you still have to pick one as the canonical instance in the module registry, otherwise you will get double instancing, so which do we make it? The point is that this is a place where package authors have to make a choice, and they should at least be able to get all the information to make it.

> It is unambiguously less performant to use import maps than not - if performance is a goal, you should be avoiding import maps like the plague and ideally inlining all your specifiers that may have initially relied on them to exact paths

Import maps offer great caching benefits for some contexts, specifically where just a few library dependencies are mapped which are updated less regularly than the rest of the application build.

All we are trying to do is set guidance on features that doesn't cause future frictions. We do have a responsibility to set the guidance on features we ship, and that's what carrying out this work involves.

Is there anything actionable here in this discussion you would like to suggest, short of pulling the section / guidance on using extensions in exports entirely? If so, does that mean you would rather package authors just throw a dice / go with their personal preference? I completely get if that is the ask.

Reviewer 2: That is the ask from me - that package authors go with their personal preference, and node offer NO recommendation whatsoever, since it's not node's place to have one here.

Author: The current iteration of the PR makes no explicit recommendation, and makes it clear this is the package authors decision either way.

<!-- YAML
added:
  - v14.13.0

>   - v12.20.0

Reviewer: Why was that line deleted?

Author: As alluded to in the original description, if we are using pattern trailers in the examples, or other later features, while stating version support to 12, it gets quite confusing to separate the specific exports features support, and we want to encourage pattern trailers mappings (`"./*.js": "./*.js"`) in examples over without (`"./*": "./*.js"` or `"./*": "./*"`) for the reasons around file extension handling.

Author: Are we able to include description metadata with these version items?

Reviewer 3: Maybe like this? https://github.com/nodejs/node/blob/611847bd88ae2af952e0bc5b810edf1765452c71/doc/api/esm.md?plain=1#L326-L331

If you want to avoid 12.20.0 appearing in the generated HTML, maybe mention it in a `description` line for 14.13.0.

Author: The way I think about it is this - patterns were experimental for a while and went through some changes, and the first non-buggy implementations should be considered the stable baseline.

And simply, Node.js 12 is not a stable patterns target because most shouldn't need to know the difference between the pattern support granularity here.

I'd rather a user doesn't use patterns to support Node.js 12, than contort to avoiding pattern trailers as somehow unstable, that's the issue this is trying to address.

Author: Yeah I suppose so.

Ok, instead I've included the extra change items here detailing when imports support for patterns was added, and when pattern trailers support was added. The example uses both of these features but since we're now explicitly showing this history, the stable baseline is clear.

value, including if it contains any `/` separators.

```js
import featureX from 'es-module-package/features/x.js';
// Loads ./node_modules/es-module-package/src/features/x.js

import featureY from 'es-module-package/features/y/y.js';
// Loads ./node_modules/es-module-package/src/features/y/y.js

import internalZ from '#internal/z.js';
// Loads ./node_modules/es-module-package/src/internal/z.js
```

Including the `"*.js"` on both sides of the mapping allows restricting which
file extensions can be resolved in the case of other files like source maps

> existing, while also retaining file extensions for the consumed subpath as is
> recommend.

Reviewer 5: ```suggestion
existing, while also retaining file extensions for the consumed subpath as its
recommended.
```

Reviewer: ```suggestion
existing, while also retaining file extensions for the consumed subpath as is
recommended.
```

value, including if it contains any `/` separators.

```js
import featureX from 'es-module-package/features/x.js';
// Loads ./node_modules/es-module-package/src/features/x.js

import featureY from 'es-module-package/features/y/y.js';
// Loads ./node_modules/es-module-package/src/features/y/y.js

import internalZ from '#internal/z.js';
// Loads ./node_modules/es-module-package/src/internal/z.js
```

Including the `"*.js"` on both sides of the mapping allows restricting which
file extensions can be resolved in the case of other files like source maps
existing, while also retaining file extensions for the consumed subpath as is

> recommended.

Reviewer 2: i don't think node should be making this recommendation at all.

Author: I've added a new dedicated section on interoperability with import maps and linked to that, this makes it an interoperability note and choice, based on being informed about the interoperability concern. Instead of saying "extensions are recommended" instead the guidance then becomes "if interoperability with import maps is desired, extensions are recommended".

As with the previous approach, a variant of this approach not requiring
conditional exports for consumers could be to add an export, e.g.
`"./module.mjs"`, to point to an all-ES module-syntax version of the package:

```json
// ./node_modules/pkg/package.json
{
  "type": "module",
  "exports": {
    ".": "./index.cjs",

>     "./module.mjs": "./index.mjs"

Reviewer 2: it's both unnecessary and imo harmful; see this thread: https://github.com/nodejs/node/pull/43375#discussion_r895292731

>     "./module.mjs": "./index.mjs"

Author: I've removed the extension for this case, that way we have some examples with and some without subpath extensions now.
